### PR TITLE
[Enhancement] Support parquet footer cache.

### PR DESCRIPTION
**Pull request by GavinMar**
Parsing parquet footer usually bring lots of CPU overhead when the parquet file has a large footer.
So, we support parquet footer cache based on object cache functions in starcache library. It can store a parsed `FileMetaData` object in the cache, and we can reuse it directly later avoid parsing the footer data repeatedlly.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 80 / 84 (95.24%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/io/cache_input_stream.cpp|1|2|50.00%|[155]|
|:large_blue_circle:|src/block_cache/starcache_wrapper.cpp|3|5|60.00%|[60, 61]|
|:large_blue_circle:|src/block_cache/block_cache.cpp|12|13|92.31%|[115]|
|:large_blue_circle:|src/exec/hdfs_scanner.h|1|1|100.00%|[]|
|:large_blue_circle:|src/connector/hive_connector.cpp|2|2|100.00%|[]|
|:large_blue_circle:|src/exec/hdfs_scanner_parquet.cpp|9|9|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/file_reader.cpp|47|47|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/metadata.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/meta_helper.h|4|4|100.00%|[]|

File: be/src/formats/parquet/metadata.h [hunk 1]
```diff
@@ -87,6 +87,11 @@ class FileMetaData {
 
     const ApplicationVersion& writer_version() const { return _writer_version; }
 
+    size_t estimate_memory() const {
+        size_t version_size = _writer_version.application_.length() + _writer_version.build_.length();
+        return sizeof(tparquet::FileMetaData) + sizeof(uint64_t) + _schema.estimate_memory() + version_size;
```
Comment (letian-jiang) [inline be/src/formats/parquet/metadata.h:92]: Is using `sizeof` accurate? On-heap memory is not counted

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 82 / 87 (94.25%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/io/cache_input_stream.cpp|1|2|50.00%|[155]|
|:large_blue_circle:|src/block_cache/starcache_wrapper.cpp|3|5|60.00%|[60, 61]|
|:large_blue_circle:|src/block_cache/block_cache.cpp|13|15|86.67%|[115, 134]|
|:large_blue_circle:|src/exec/hdfs_scanner.h|1|1|100.00%|[]|
|:large_blue_circle:|src/connector/hive_connector.cpp|2|2|100.00%|[]|
|:large_blue_circle:|src/exec/hdfs_scanner_parquet.cpp|11|11|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/file_reader.cpp|46|46|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/metadata.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/meta_helper.h|4|4|100.00%|[]|

File: be/src/formats/parquet/metadata.h (hunk 1 above)
Comment (GavinMar) [reply be/src/formats/parquet/metadata.h:92]: > Is using `sizeof` accurate? On-heap memory is not counted

Not accurate, and in fact it is difficult to calculate the accurate memory of this class objects. So we add some comments in the related functions. In the other hand, this value is only used for simple memory estimation and does not need to be completely accurate.

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 82 / 87 (94.25%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/io/cache_input_stream.cpp|1|2|50.00%|[155]|
|:large_blue_circle:|src/block_cache/starcache_wrapper.cpp|3|5|60.00%|[60, 61]|
|:large_blue_circle:|src/block_cache/block_cache.cpp|13|15|86.67%|[115, 134]|
|:large_blue_circle:|src/exec/hdfs_scanner.h|1|1|100.00%|[]|
|:large_blue_circle:|src/connector/hive_connector.cpp|2|2|100.00%|[]|
|:large_blue_circle:|src/exec/hdfs_scanner_parquet.cpp|11|11|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/file_reader.cpp|46|46|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/metadata.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/meta_helper.h|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 100 / 130 (76.92%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/block_cache/starcache_wrapper.cpp|9|21|42.86%|[64, 65, 66, 67, 68, 69, 71, 92, 93, 94, 95, 97]|
|:large_blue_circle:|src/connector/hive_connector.cpp|2|3|66.67%|[106]|
|:large_blue_circle:|src/io/cache_input_stream.cpp|2|3|66.67%|[273]|
|:large_blue_circle:|src/formats/parquet/file_reader.cpp|37|52|71.15%|[51, 52, 53, 54, 55, 56, 57, 58, 63, 64, 65, 67, 68, 75, 76]|
|:large_blue_circle:|src/block_cache/block_cache.cpp|15|16|93.75%|[117]|
|:large_blue_circle:|src/exec/hdfs_scanner_parquet.cpp|16|16|100.00%|[]|
|:large_blue_circle:|src/exec/hdfs_scanner.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/schema.h|9|9|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/metadata.cpp|2|2|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/metadata.h|3|3|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/meta_helper.h|4|4|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 2 / 2 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/qe/SessionVariable.java|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 100 / 130 (76.92%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/block_cache/starcache_wrapper.cpp|9|21|42.86%|[64, 65, 66, 67, 68, 69, 71, 92, 93, 94, 95, 97]|
|:large_blue_circle:|src/connector/hive_connector.cpp|2|3|66.67%|[106]|
|:large_blue_circle:|src/io/cache_input_stream.cpp|2|3|66.67%|[273]|
|:large_blue_circle:|src/formats/parquet/file_reader.cpp|37|52|71.15%|[51, 52, 53, 54, 55, 56, 57, 58, 63, 64, 65, 67, 68, 75, 76]|
|:large_blue_circle:|src/block_cache/block_cache.cpp|15|16|93.75%|[117]|
|:large_blue_circle:|src/exec/hdfs_scanner_parquet.cpp|16|16|100.00%|[]|
|:large_blue_circle:|src/exec/hdfs_scanner.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/schema.h|9|9|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/metadata.cpp|2|2|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/metadata.h|3|3|100.00%|[]|
|:large_blue_circle:|src/formats/parquet/meta_helper.h|4|4|100.00%|[]|

**Review by letian-jiang — APPROVED** [review]

**Review by Youngwb — DISMISSED** [review]

**Review by Youngwb — DISMISSED** [review]

**Review by stephen-shelby — DISMISSED** [review]

**Review by stephen-shelby — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Review by wangruin — APPROVED** [review]

**Pull request merged**